### ISSUE TITLE
API Design: define entities by archetype directly

### ISSUE THREAD

**Issue by Avokadoen** [Avokadoen/ecez#29]
Building on the features made in the comptime branch we can design the API to utilize compile time more and enabling these features to be merged in master. Merging comptime branch in master will resolve issue #27 

The user will have to define archetypes i.e:

```zig
const Player = struct {
   // Note variable name is not important here
   hp: Health,
   pos: Position,
   vel:  Velocity,
};

const StaticObject = {
   Pos: Position,
};
```

These archetypes are then stored in a comptime hash map

The API will not allow dynamically adding or removing components from an entity, but changing the entity type will be valid in any direction (or maybe this should be restricted?)

This will like mentioned allow the library to move a lot more over to compile time and hopefully make planned queries mostly compile time based as well and add the functionality for type queries i.e ``give me all Players``

It will also make entity construction very fast since the user will always specify the full component list at creation i.e

```zig
world.setEntityType(entity, Player{...});
```

**Comment by Avokadoen** [comment]
each archetype should implement an [interface](https://zig.news/david_vanderson/faster-interface-style-2b12) 
we can then have an array of these interfaces and then we can make some operations O(1) i.e ``hasComponent`` and ``getComponent`` etc...

**Comment by Avokadoen** [comment]
The restriction of only allowing predefined archetypes can be removed by using the interface idea 

Queries can be processed on compile time using comptime and itnerfaces can be used to do partial type erasure needed to combine different archetypes!

**Comment by Avokadoen** [comment]
Resolved by PR #30
